File: Array/liuchang/day25/day25.java
public class day25 {
    public static class Node {
        int data;
        Node next;

        public Node(int data) {
            this.data = data;
        }

        public void setNext(Node next) {
            this.next = next;
        }

        public int hasCycle(Node head) {
            int count = -1;
            Node slowNode = head;
            Node fastNode = head;
            if (slowNode.next == null || fastNode.next == null) {
                return -1;
            }
            for (; ; ) {
                slowNode = slowNode.next;
                fastNode = fastNode.next.next;
                if (slowNode == fastNode) {
                    return count;
                } else if (fastNode == null) {
                    return count;
                } else {
                    count++;
                }
            }

        }

    }

    public static void main(String[] args) {
        Node node1 = new Node(3);
//        Node node2 = new Node(2);
//        node1.setNext(node2);
//        node2.setNext(node1);
//        Node node3 = new Node(0);
//        node2.setNext(node3);
//        Node node4 = new Node(4);
//        node3.setNext(node4);
//        node4.setNext(node2);
        System.out.println("=====" + node1.hasCycle(node1));
    }

}

File: Array/liuchang/day25/day25.kt
package net.study.kotlintest

/**
 * @author       刘畅
 * @createdate   2020/6/24
 * @describe
 */
class day25 {
    class Node {
        var data = -1;
        var next: Node? = null;

        constructor(data: Int) {
            this.data = data
        }


        fun hasCycle(head: Node): Int {
            var count = -1;
            var slowNode = head;
            var fastNode = head;
            if (head.next == null) {
                return count;
            }
            for (i in 0 until Int.MAX_VALUE) {

                slowNode = slowNode.next!!;
                if (fastNode.next!!.next != null) {
                    fastNode = fastNode.next!!.next!!;
                } else {
                    return -1;
                }
                if (slowNode == fastNode) {
                    return count;
                } else if (fastNode == null) {
                    return count;
                } else {
                    count++;
                }
            }
            return count;
        }

    }
    fun main(args: Array<String>) {
        var node1 = Node(3);
        var node2 = Node(2);
        node1.next = node2;
        var node3 = Node(0);
        node2.next = node3;
        var node4 = Node(4);
        node3.next = node4;
//        node4.setNext(node2);
        System.out.println("=====" + node1.hasCycle(node1));
    }
}